src/python/gudhi/wasserstein/wasserstein.py

```diff
@@ -173,9 +173,40 @@ def wasserstein_distance(X, Y, matching=False, order=2., internal_p=2., enable_a
         return ep.concatenate(dists).norms.lp(order).raw
         # We can also concatenate the 3 vectors to compute just one norm.
 
-    # Comptuation of the otcost using the ot.emd2 library.
+    # Comptuation of the otcost using the ot.emd2 library, and matching/transport plan using ot.emd.
     # Note: it is the Wasserstein distance to the power q.
     # The default numItermax=100000 is not sufficient for some examples with 5000 points, what is a good value?
-    ot_cost = ot.emd2(a, b, M, numItermax=2000000)
+    if reg == 0.:
+        if matching:
+            P = ot.emd(a=a,b=b,M=M, numItermax=2000000)
+            ot_cost = np.sum(np.multiply(P,M))
+            P[-1, -1] = 0  # Remove matching corresponding to the diagonal
+            match = np.argwhere(P)
+            # Now we turn to -1 points encoding the diagonal
+            match[:,0][match[:,0] >= n] = -1
+            match[:,1][match[:,1] >= m] = -1
+            return ot_cost ** (1./order) , match
+        else:
+            ot_cost = ot.emd2(a, b, M, numItermax=2000000)
+
+    elif reg > 0:
+        np.seterr(divide='raise')  # For a better handling of numerical issues due to reg being too low
+        try:
+            if matching:
+                P = ot.sinkhorn(a=a, b=b, M=M, reg=reg)
+                ot_cost = np.sum(np.multiply(P, M))
+                return ot_cost ** (1./order), P
+            else:
+                ot_cost = ot.sinkhorn2(a, b, M, reg=reg)[0]
+                return ot_cost ** (1. / order)
+        except FloatingPointError:
+            raise FloatingPointError('Error when computing regularized wasserstein distance (Sinkhorn loop).' \
+                                     'This is probably due to reg being too low. Try' \
+                                     'to increase reg or set it to 0 for exact computation')
+    else:
+        raise ValueError('Negative value for reg')
+        # It seems ot.sinkhorn2 returns something when reg<0, which is reasonnable but probably unexpected
```

Comment:
Interesting. If you put a very large negative reg, what should it converge to, should it stay at the same value as reg=0 or send every point to the diagonal?

Reply:
If reg is super negative, it seems you obtain a similar behavior as reg being super positive, that is you obtain the "trivial transport plan", where each point is transported uniformly to all other points and the diagonal. 
The reg parameter only appears through the matrix K = exp( - M / reg), (term-wise exponentiation) where M is the pairwise cost matrix (including the diagonal), and then the only requirement you have about this matrix is to have positive entries (which is the case, even with reg < 0). Using reg < 0 is somewhat equivalent to use negative distances as your transport cost.

Reply:
I see. Thanks.